### ISSUE TITLE
Signature didn't verify. Have the contents been tampered with?

### ISSUE THREAD

**Issue by pierredepascale** [Sustainsys/Saml2#1302]
# Problem login with Schibboleth IDP

We are using the sustainsys.saml2 library to provide SAML login in our application. When we use our WebApp against a Schibboleth based IDP we get the following exception "Signature didn't verify.". We know there is one signing key in the IDP that
comes from loading the metadata.

# Additional Info

The AuthNRequest is sent to the IDP. When we receive the HttpPost response from the IDP we get the exception:

[2021-08-12 16:36:23,642] [ERROR] [ 16] [] [x0dst5vcgv02ay5hpncajlyu] [172.24.52.108] [...] Sustainsys.Saml2.Exceptions.InvalidSignatureException: Signature didn't verify. Have the contents been tampered with?
   bei Sustainsys.Saml2.XmlHelpers.VerifySignature(IEnumerable`1 signingKeys, SignedXml signedXml, XmlElement signatureElement, Boolean validateCertificate)
   bei Sustainsys.Saml2.XmlHelpers.IsSignedByAny(XmlElement xmlElement, IEnumerable`1 signingKeys, Boolean validateCertificate, String minimumSigningAlgorithm)
   bei Sustainsys.Saml2.Saml2P.Saml2Response.ValidateSignature(IOptions options, IdentityProvider idp)
   bei Sustainsys.Saml2.Saml2P.Saml2Response.<CreateClaims>d__66.MoveNext()
   bei System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   bei System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   bei Sustainsys.Saml2.Saml2P.Saml2Response.GetClaims(IOptions options, IDictionary`2 relayData)
   bei Sustainsys.Saml2.WebSso.AcsCommand.ProcessResponse(IOptions options, Saml2Response samlResponse, StoredRequestState storedRequestState, IdentityProvider identityProvider, String relayState)
   bei Sustainsys.Saml2.WebSso.AcsCommand.Run(HttpRequestData request, IOptions options)
   bei Edoc.Panels.WebApi.Controllers.EdocSaml2Controller.Acs() in C:\...
.
We tested successfully our application with 2 other IDPs (home made one and KeyCloak).

The difference we see between the various tests are that the `<ds:SignatureValue>` is on one line for success and multiple line for failure. Here is the XML response that fails the verification:

```xml
<saml2p:Response xmlns:saml2p="urn:oasis:names:tc:SAML:2.0:protocol" xmlns:xs="http://www.w3.org/2001/XMLSchema" Destination="https://esc9100-01.isc.ejpd.admin.ch/EdocSaml2/Acs" ID="_f5aca476-1eb1-4ed8-970f-02227dfc53bf" InResponseTo="idc51b5b84db314f1ea8bf653707b83c9e" IssueInstant="2021-08-12T14:36:40.881Z" Version="2.0"><saml2:Issuer xmlns:saml2="urn:oasis:names:tc:SAML:2.0:assertion">https://syspvm0009-int.isc.ejpd.admin.ch:8543/idp/saml2/metadata</saml2:Issuer><ds:Signature xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
<ds:SignedInfo>
<ds:CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"></ds:CanonicalizationMethod>
<ds:SignatureMethod Algorithm="http://www.w3.org/2001/04/xmldsig-more#rsa-sha256"></ds:SignatureMethod>
<ds:Reference URI="#_f5aca476-1eb1-4ed8-970f-02227dfc53bf">
<ds:Transforms>
<ds:Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"></ds:Transform>
<ds:Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"><ec:InclusiveNamespaces xmlns:ec="http://www.w3.org/2001/10/xml-exc-c14n#" PrefixList="xs"></ec:InclusiveNamespaces></ds:Transform>
</ds:Transforms>
<ds:DigestMethod Algorithm="http://www.w3.org/2001/04/xmlenc#sha256"></ds:DigestMethod>
<ds:DigestValue>modrWqqwj//Iyb0Oh6q4ZBEirm7ttiCkK5fXWMnXnQg=</ds:DigestValue>
</ds:Reference>
</ds:SignedInfo>
<ds:SignatureValue>lkDDWHxM8KiQVepW8edCEn/2jJXIoFN+GTmpJBSvg2f3EXgUdnW0I2snzR/b/O4Jsj7v+qEgMDdR
kmS4A+iRJ0F2FqZQsDbevqa1UVg8B/YIvwz061mvdN5r4d2lZtgYFljGoSAz0dTZBQX2Whz2Qdg8
LGjF3ubm7XMmC+2wKoT+s+snNJzpzuUcWJHBksQJUmiPtl9kQarI2oTcjVqX4xnGsgE6fu1WZFgl
5Cm/PlAV8DzdMcIPuzM77AV3WWISBVGbewomMfJ0j3O5xcOrXV4IhaVMA85F+GXXua/DFict46pn
ScFqXwmMX/rPDSv8uKTJycR7U5qJNjav63rZjA==</ds:SignatureValue>
<ds:KeyInfo><ds:X509Data><ds:X509Certificate>MIIHhTCCBW2gAwIBAgIQGCY+6z/aHNWLfYqOJjzuwjANBgkqhkiG9w0BAQsFADB9MQswCQYDVQQG

...

dyxHfkP6BQeryae2jwkgtsZa6SUx96m</ds:X509Certificate></ds:X509Data></ds:KeyInfo></ds:Signature><saml2p:Status><saml2p:StatusCode Value="urn:oasis:names:tc:SAML:2.0:status:Success"></saml2p:StatusCode></saml2p:Status><saml2:Assertion xmlns:saml2="urn:oasis:names:tc:SAML:2.0:assertion" xmlns:xs="http://www.w3.org/2001/XMLSchema" ID="_deefb818-7d5e-4833-9713-90797f40110d" IssueInstant="2021-08-12T14:36:40.881Z" Version="2.0"><saml2:Issuer>https://syspvm0009-int.isc.ejpd.admin.ch:8543/idp/saml2/metadata</saml2:Issuer><ds:Signature xmlns:ds="http://www.w3.org/2000/09/xmldsig#">
<ds:SignedInfo>
<ds:CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"></ds:CanonicalizationMethod>
<ds:SignatureMethod Algorithm="http://www.w3.org/2001/04/xmldsig-more#rsa-sha256"></ds:SignatureMethod>
<ds:Reference URI="#_deefb818-7d5e-4833-9713-90797f40110d">
<ds:Transforms>
<ds:Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"></ds:Transform>
<ds:Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"><ec:InclusiveNamespaces xmlns:ec="http://www.w3.org/2001/10/xml-exc-c14n#" PrefixList="xs"></ec:InclusiveNamespaces></ds:Transform>
</ds:Transforms>
<ds:DigestMethod Algorithm="http://www.w3.org/2001/04/xmlenc#sha256"></ds:DigestMethod>
<ds:DigestValue>b0+hLhF+Z94KF9SibxpZcXdAPzPTu2vu5/lwA9Tj9bg=</ds:DigestValue>
</ds:Reference>
</ds:SignedInfo>
<ds:SignatureValue>Kh3AoE0/nvz892ceybGa54DFc07CmoFEsMWu6yocTQlDSleysV53UCcdU5xqSE9uA99vVftNF6ec
Xt4HnqDE5FfPMnRB8xXfyTzIbyRRDmMrYW4+3XOi5sqpyVfe0gsfTIyYpQLlkudArkd/xaPqIPk7
vjzU0d8dIwZD/XuSAXDxaPDpiXjLvAXlxRdRHxQ9L0HkTVoHmsmcB+pLAZE450c3709ghYVkXRjy
HHsnRaYlHIXKeppmXYEeKqAiPKvXcn0W29QyC6r1DZmwMpnyKFMvZ4ezOw/hgKBLoFR+VM4GeJdy
R1W4J5wINhw4FQyPnQ/rIgnECSm8+lJB/qVW0A==</ds:SignatureValue>
<ds:KeyInfo><ds:X509Data><ds:X509Certificate>MIIHhTC
... 
a6SUx96m</ds:X509Certificate></ds:X509Data></ds:KeyInfo></ds:Signature><saml2:Subject><saml2:NameID Format="urn:oasis:names:tc:SAML:2.0:nameid-format:transient" NameQualifier="https:// ... saml2/metadata" SPNameQualifier="https://e ... Saml2">AAdzZWN ... 1FLGg=</saml2:NameID><saml2:SubjectConfirmation Method="urn:oasis:names:tc:SAML:2.0:cm:bearer"><saml2:SubjectConfirmationData Address="10.0.0.1" InResponseTo="idc51b5b84db314f1ea8bf653707b83c9e" NotOnOrAfter="2021-08-12T14:41:40.881Z" Recipient="https://e ... Saml2/Acs"></saml2:SubjectConfirmationData></saml2:SubjectConfirmation></saml2:Subject><saml2:Conditions NotBefore="2021-08-12T11:31:40.881Z" NotOnOrAfter="2021-08-13T08:36:40.881Z"><saml2:AudienceRestriction><saml2:Audience>https://e ... Saml2</saml2:Audience></saml2:AudienceRestriction></saml2:Conditions><saml2:AuthnStatement AuthnInstant="2021-08-12T14:36:40.881Z" SessionIndex="C4C0333C3CD2B2D8A59F0EC12A21C8C0"><saml2:SubjectLocality Address="10.0.0.1"></saml2:SubjectLocality><saml2:AuthnContext><saml2:AuthnContextClassRef>urn:oasis:names:tc:SAML:2.0:ac:classes:X509</saml2:AuthnContextClassRef></saml2:AuthnContext></saml2:AuthnStatement><saml2:AttributeStatement><saml2:Attribute FriendlyName="uid" Name="urn:oid:0.9.2342.19200300.100.1.1" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri"><saml2:AttributeValue xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:type="xs:anyType">83006060</saml2:AttributeValue></saml2:Attribute><saml2:Attribute FriendlyName="userCertificate" Name="urn:oid:2.5.4.36" NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:uri"><saml2:AttributeValue xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:type="xs:anyType">CN= ..., C=JP</saml2:AttributeValue></saml2:Attribute></saml2:AttributeStatement></saml2:Assertion></saml2p:Response>
```

One example where the `SignatureValue` is on one line:

```xml
<ds:DigestMethod Algorithm="http://www.w3.org/2001/04/xmlenc#sha256"></ds:DigestMethod>
<ds:DigestValue>b0+hLhF+Z94KF9SibxpZcXdAPzPTu2vu5/lwA9Tj9bg=</ds:DigestValue>
</ds:Reference>
</ds:SignedInfo>
<ds:SignatureValue>Kh3AoE0/nvz892ceybGa54DFc07CmoFEsMWu6yocTQlDSleysV53UCcdU5xqSE9uA99vVftNF6ec&#xD;Xt4HnqDE5FfPMnRB8xXfyTzIbyRRDmMrYW4+3XOi5sqpyVfe0gsfTIyYpQLlkudArkd/xaPqIPk7&#xD;vjzU0d8dIwZD/XuSAXDxaPDpiXjLvAXlxRdRHxQ9L0HkTVoHmsmcB+pLAZE450c3709ghYVkXRjy&#xD;HHsnRaYlHIXKeppmXYEeKqAiPKvXcn0W29QyC6r1DZmwMpnyKFMvZ4ezOw/hgKBLoFR+VM4GeJdy&#xD;R1W4J5wINhw4FQyPnQ/rIgnECSm8+lJB/qVW0A==</ds:SignatureValue>
<ds:KeyInfo><ds:X509Data><ds:X509Certificate>MIIHh
```

# Setup

Nuget package version=2.8.0 latest
dotnet compile= net472
dotnet runtime/installed=net472
IIS Version 10.1809
Asp.Net MVC 5.2.7

Why the signature verification failing in our case ?

**Comment by immilev** [comment]
Ditto. I have had the same issue - I suspect Sustainsys2 does not handle the line-breaks in SignatureValue correctly. My Response is generated by OpenSAML by Shibboleth. Any ideas why? @AndersAbel

**Comment by tbudai** [comment]
My view on this is that this is an IDP side issue.

As documented [here](https://www.tabnine.com/web/assistant/code/rs/5c6564811095a500014953a8#L232):
Due to the change to sha256 the hash length increased and caused XmlSec to add linebreaks to the generated digest, which resulted in an invalid signature.

From my POV nothing to do on Sustainsys/SAML2 side.

**Comment by tbudai** [comment]
SignatureValue is of type base64Binary.
As per [spec](https://www.w3.org/TR/xmlschema11-2/#base64Binary):

`The ·lexical representations· of base64Binary values are limited to the 65 characters of the Base64 Alphabet defined in [RFC 3548], i.e., a-z, A-Z, 0-9, the plus sign (+), the forward slash (/) and the equal sign (=), together with the space character (#x20). No other characters are allowed.`

That's why `&#xD;` is not handled on client side and remains within SignatureValue.

**Comment by AndersAbel** [comment]
Thanks @tbudai for your investigation. Based on that, I agree that this is an Idp issue. I'm closing this as invalid/wontfix.

**Comment by Shumiii** [comment]
> I see SignatureValue contains `&#xD;` Maybe the issue is related to https://shibboleth.atlassian.net/browse/OSJ-267
> 
> So a solution might be to have Shibboleth IDP side setting `org.apache.xml.security.ignoreLineBreaks=true`

I too am having this issue today. I have SAML responses coming in from Azure AD. Most of the assertions work fine but a few specific accounts have this same verify signature issue. @tbudai I hunted for the line break you mentioned (&#xD) in the erroneous signature value but could not see one. I then looked into the OP's signature value and do not see one there either although you said you see one. What am I missing?

_For reference here is a signature value that errors:_
G9jCJJN6aDmJqhciN1nF3lvkq2+GCaPYegTsxiVTtyqwrFTyQYKJggyVdDAiuG3mqxhZsE2Lg+8CXaafOLRfU5B3IdU8n625JCSAkXfEu8OcFu9NuOlUVFSlJc5iGBEP9uIRmvimZqljz0CmoJgun1hB38NB8aWEZTmUVRdiMUz4dgYUzTpkuMhhfGBEPNu7YfvSXgzYdLAfFrt6yJS+Hjbp7fiP34ulS7qT2VFTN1Ga/rPN/WrnUXuq3y2eceh3X4vTw/owkbtHKYwLOdhH02pYHL9oUg/Na6wSc+QJeu6ECQbICvnB7CwnuU+5BcDDI5MZuC1ghXUvoj1gS+nQEw==

_And here is the signature value that works fine:_
aNFs933L0y1Qko5d55/Ohmys5ucwAGPTPXQ9jhCdJ/CUXvERdfI7zGGDLm1sT0vXLJRKKU/8vWhdyNx8YPIwdEzqVeC5xRZVDA0BI3JbSEs3c8sZ9+ap/eJrMy5z7j/7Bhm2YcJUeF51/0jaHd9GSvWt9t3nDuDwO6bjfCQ63qUUK1jFLIOBeAGuQjz55aUjgIxU9xe+qOAuf+X1XPvwDMuf90H2kzuRg3cSCDbwpSrBABBEGg5V0CFhds5T3Ldi4fWMmzCjOT6rO85PR1D4yegVtg5G48r4KBC0fKJtkGxaXkWlJQZaC13SVonPmsjB7LbuIqm7AI/a9nK2ZA9tNA==
